econml/_cate_estimator.py

```diff
@@ -742,11 +820,47 @@ def _expand_treatments(self, X=None, *Ts):
                 T = np.full((n_rows,) + self._d_t_in, T)
 
             if self.transformer:
-                T = self.transformer.transform(T)
+                T = self.transformer.transform(reshape(T, (-1, 1)))
             outTs.append(T)
 
         return (X,) + tuple(outTs)
 
+    def _set_transformed_treatment_names(self):
+        """Works with sklearn OHEs"""
+        if hasattr(self, "_input_names"):
+            self._input_names["treatment_names"] = self.transformer.get_feature_names(
+                self._input_names["treatment_names"]).tolist()
+
+    def __setattr__(self, name, value):
```

Comment:
This seems like the wrong approach to me... Among other things it seems like it would break serialization/deserialization, but also it only handles the repetition in setting categories but not in defining the transformer from that.  I think it would be more natural to override fit itself, never modify the user-set categories, and just define a local variable that handles the list wrapping, and then set transformer based on that, and have all subclasses call super.fit (except that _OrthoLearner probably needs to override it to deal with the fact that continuous treatments don’t need transformers).

Reply:
So I don't agree with this. Here is why:

1. I tested it and it does not break serialization. Classes with overridden setters can be pickled.
2. The transformer is an OHE that is defined at `fit` time with whatever `self.categories` is defined at that point. The alternative is to define a wrapped `categories` in every `fit`, thus duplicating code everywhere and having the same issue as @vasilismsr mentioned: there is boilerplate code that is repeated everywhere despite the fact that it belongs to `TreatmentExpansionMixin` and that makes the code hard to follow and it makes it hard for others to contribute.
3. I am against overriding `fit`. `fit` is already very complicated due to the inheritance and the wrapping. The ORF already has that issue and needs to override `fit` everywhere and losing some of the logic implemented in the `BaseCateEstimator`. I am open to other alternatives, but I think this is a recipe for more issues down the line. 

Since the `categories` attribute is defined some time before `fit`, we could move the transformer definition in the `prefit`, however that will have some implication there that there is a `categories` attribute in the child class that is not mentioned in `TreatmentExpansionMixin` and referencing attributes that might be in a child makes me uneasy. Furthermore, the transformer definition is tied to whether there is a discrete treatment which is again an attribute of the child class and not of the `TreatmentExpansionMixin`. This is an issue for both the `_OrthoLearner` and the `DMLOrthoForest` classes.